security/pkg/pki/custom/ca.go

```diff
@@ -0,0 +1,181 @@
+		return nil, fmt.Errorf("verify CSR Identities failed: %v", err)
+	}
+	cLog.Debugf("forwarding Workload's CSR to Custom CA server...")
+	resp, err := c.signingAPIClient.CreateCertificate(ctx, req)
+	certChain := resp.GetCertChain()
+	if err != nil {
+		cLog.Errorf("cannot call CreateCertificate from Custom CA: %v", err)
+		return nil, fmt.Errorf("cannot call CreateCertificate from Custom CA: %v", err)
+	}
+
+	var responseCertChains []string
+	certChainWithoutRoot := certChain[:len(certChain)-1]
+	for _, cert := range certChainWithoutRoot {
+		parsedCert, err := validateAndParseCert(cert)
+		if err != nil {
+			cLog.Errorf("response certificate from Custom CA is invalid: %v", err)
+			return nil, fmt.Errorf("response certificate from Custom CA is invalid: %v", err)
+		}
+		responseCertChains = append(responseCertChains, parsedCert)
+	}
+
+	// Append current roots: Custom CA's root CA and Self signing's root CA
+	rootCertBytes := c.keyCertBundle.GetRootCertPem()
+	responseCertChains = append(responseCertChains, string(rootCertBytes))
+
+	return &pb.IstioCertificateResponse{
+		CertChain: responseCertChains,
+	}, nil
+}
+
+func validateAndParseCert(cert string) (string, error) {
+	defer responseTimeTrack(time.Now(), "validateAndParseCert")
+	certBytes, _ := pem.Decode([]byte(cert))
+
+	if certBytes == nil {
+		return "", fmt.Errorf("decode cert is failed, invalid certificate: %v", cert)
+	}
+
+	block := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes.Bytes,
+	}
+
+	c := pem.EncodeToMemory(block)
+	return string(c), nil
+}
+
+func validateCSR(identities []string, csrPEM string) error {
+	if len(identities) == 0 {
+		return nil
+	}
+
+	csrParsed, _ := pem.Decode([]byte(csrPEM))
+	if csrParsed == nil {
+		return fmt.Errorf("decode csr failed - empty pem: %v", csrPEM)
+	}
+	parsedCSR, err := x509.ParseCertificateRequest(csrParsed.Bytes)
+	if err != nil {
+		return fmt.Errorf("cannot parse CSR: %v", err)
+	}
+
+	if len(parsedCSR.Subject.CommonName) != 0 {
+		return fmt.Errorf("received unsupported CSR, contains Subject.CommonName: %#v", parsedCSR.Subject.CommonName)
+	}
+
+	if len(parsedCSR.DNSNames) != 0 {
+		return fmt.Errorf("received unsupported CSR, contains DNSNames: %v", parsedCSR.DNSNames)
+	}
+
+	if len(parsedCSR.IPAddresses) != 0 {
+		return fmt.Errorf("received unsupported CSR, contains IPAddresses: %v", parsedCSR.IPAddresses)
+	}
+
+	if len(parsedCSR.EmailAddresses) != 0 {
+		return fmt.Errorf("received unsupported CSR, contains EmailAddresses: %v", parsedCSR.EmailAddresses)
+	}
+
+	idsFromCSR, err := util.ExtractIDs(parsedCSR.Extensions)
+	if err != nil {
+		return fmt.Errorf("cannot extract identities from CSR: %v", err)
+	}
+
+	if len(idsFromCSR) == 0 {
+		return fmt.Errorf("missing spiffe identities in CSR")
+	}
+	identitiesAsStr := strings.Join(identities, ",")
+
+	unmatched := false
+	for _, csrID := range idsFromCSR {
+		if !strings.Contains(identitiesAsStr, csrID) {
+			unmatched = true
+		}
+	}
```

Comment:
I think this is not quite right yet.

I can have an identity of `istiod-foo`, but successfully submit a CSR with `istiod` for example.

Reply:
identities are SPPIFEE Format

Reply:
Sure, but that doesn't change the issue:

Identity: `spiffe://admin/ns/admin/sa/admin-bookinfo`
Accepted CSR: `spiffe://admin/ns/admin/sa/admin`

This would allow a  signed certificate for any identity which is a sub-string match of the real identity. This would allow an attacker to attain a different identity (unacceptable), and try and leverage to either intercept or privilege escalate/impersonate.

Rather than testing for a substring match on the concatenated identities, I recommend string matching on a sorted list of both. Something like below:

```go
// identitiesMatch will ensure that two lists of identities given from the
// request context, and those parsed from the CSR, match
func identitiesMatch(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}

	aa := make([]string, len(a))
	bb := make([]string, len(b))

	copy(aa, a)
	copy(bb, b)

	sort.Strings(aa)
	sort.Strings(bb)

	for i, v := range aa {
		if bb[i] != v {
			return false
		}
	}

	return true
}
```

Reply:
Thanks @JoshVanL